File: apps/server/tests/factories.py
"""Shared test-data factories for the craftcontrol backend test suite.

Use these builders instead of constructing raw dicts inline so that envelope
and snapshot shapes stay consistent when the schema evolves.
"""
from __future__ import annotations

from typing import Any

from controlplane.audit.model import AuditRecord
from controlplane.core.events import Event
from controlplane.operations.lifecycle import (
    OperationStage,
    OperationState,
    StageRecord,
    StageResult,
    ServerOperation,
)
from controlplane.telemetry.installer import TelemetryPackStatus


def telemetry_envelope(
    event_type: str = "blocks.changed",
    sequence: int = 1,
    player: str | None = "VonCrush",
    data: dict | None = None,
    timestamp: int = 1,
) -> dict:
    """Return a well-formed telemetry envelope dict.

    Args:
        event_type: The event type string (e.g. ``"blocks.changed"``).
        sequence: Monotonically increasing sequence number from the pack.
        player: Player name to embed in ``{"name": player}``; pass ``None``
            for server-level events that carry no player context.
        data: Payload dict. Defaults to an empty dict.
        timestamp: Unix timestamp. Defaults to ``1``.

    Returns:
        A dict matching the telemetry envelope schema v1.
    """
    return {
        "schema": 1,
        "sequence": sequence,
        "type": event_type,
        "timestamp": timestamp,
        "player": {"name": player} if player is not None else None,
        "data": data if data is not None else {},
    }


def player_snapshot(
    name: str = "VonCrush",
    online: bool = True,
    xuid: str = "999",
) -> dict:
    """Return a minimal player-snapshot dict suitable for seeding tests.

    Args:
        name: Display name.
        online: Whether the player is currently connected.
        xuid: Internal Xbox User ID (never exposed externally).

    Returns:
        A dict with ``name``, ``online``, and ``xuid`` keys.
    """
    return {"name": name, "online": online, "xuid": xuid}


def audit_record(
    id: int = 1,
    occurred_at: float = 1.0,
    actor: str | None = "alice",
    action: str = "auth.login",
    target: str | None = "alice",
    result: str = "success",
    metadata: dict[str, Any] | None = None,
) -> AuditRecord:
    return AuditRecord(
        id=id,
        occurred_at=occurred_at,
        actor=actor,
        action=action,
        target=target,
        result=result,
        metadata=metadata if metadata is not None else {},
    )


def event(
    id: int = 1,
    topic: str = "server.status",
    timestamp: float = 1.0,
    source: str = "test",
    payload: dict[str, Any] | None = None,
) -> Event:
    return Event(
        id=id,
        topic=topic,
        timestamp=timestamp,
        source=source,
        payload=payload if payload is not None else {},
    )


def stage_record(
    stage: OperationStage = OperationStage.REVIEW,
    result: StageResult = StageResult.PENDING,
    started_at: float | None = None,
    completed_at: float | None = None,
    evidence: dict[str, Any] | None = None,
    error: str | None = None,
) -> StageRecord:
    return StageRecord(
        stage=stage,
        result=result,
        started_at=started_at,
        completed_at=completed_at,
        evidence=evidence if evidence is not None else {},
        error=error,
    )


def server_operation(
    operation_id: str = "op-1",
    server_id: str = "default",
    requested_changes: dict[str, Any] | None = None,
    state: OperationState = OperationState.PENDING,
    stages: list[StageRecord] | None = None,
    created_at: float = 1.0,
    updated_at: float = 1.0,
    completed_at: float | None = None,
    terminal_error: str | None = None,
) -> ServerOperation:
    return ServerOperation(
        operation_id=operation_id,
        server_id=server_id,
        requested_changes=requested_changes if requested_changes is not None else {},
        state=state,
        stages=stages if stages is not None else [StageRecord(stage=s) for s in OperationStage.ordered()],
        created_at=created_at,
        updated_at=updated_at,
        completed_at=completed_at,
        terminal_error=terminal_error,
    )


def telemetry_pack_status(
    world: str = "Bedrock level",
    source_version: str = "1.0.0",
    installed_version: str | None = "1.0.0",
    enabled_version: str | None = "1.0.0",
    installed: bool = True,
    enabled: bool = True,
    upgrade_available: bool = False,
    legacy_directory: bool = False,
    restart_required: bool | None = False,
    installed_updated_at: float | None = 1.0,
) -> TelemetryPackStatus:
    return TelemetryPackStatus(
        world=world,
        source_version=source_version,
        installed_version=installed_version,
        enabled_version=enabled_version,
        installed=installed,
        enabled=enabled,
        upgrade_available=upgrade_available,
        legacy_directory=legacy_directory,
        restart_required=restart_required,
        installed_updated_at=installed_updated_at,
    )

File: apps/client/tests/factories.js
/**
 * Domain object builders for client tests.
 *
 * Each factory returns a minimal valid API-shaped object with sensible
 * defaults. Pass an overrides map to customise individual fields.
 * Builders here mirror the domain contracts defined in the backend —
 * keep them in sync when the API shapes change.
 */

export function makeOperation(overrides = {}) {
  return {
    operation_id: "op-1",
    server_id: "default",
    state: "running",
    requested_changes: { SERVER_NAME: "MyWorld" },
    stages: [
      { stage: "review",        result: "completed", started_at: 1700000000, completed_at: 1700000001, evidence: {}, error: null },
      { stage: "backup_verify", result: "running",   started_at: 1700000002, completed_at: null,       evidence: { backup_path: "/backups/world" }, error: null },
      { stage: "prepare",       result: "pending",   started_at: null,       completed_at: null,       evidence: {}, error: null },
      { stage: "restart",       result: "pending",   started_at: null,       completed_at: null,       evidence: {}, error: null },
      { stage: "health_wait",   result: "pending",   started_at: null,       completed_at: null,       evidence: {}, error: null },
      { stage: "verify",        result: "pending",   started_at: null,       completed_at: null,       evidence: {}, error: null },
      { stage: "confirm",       result: "pending",   started_at: null,       completed_at: null,       evidence: {}, error: null },
    ],
    created_at: 1700000000,
    updated_at: 1700000002,
    completed_at: null,
    terminal_error: null,
    observation: {},
    correlation_id: null,
    ...overrides,
  };
}

export function makePlayer(overrides = {}) {
  return {
    id: "abc123",
    name: "Alice",
    online: false,
    deaths_count: 2,
    total_play_seconds: 3600,
    operator: false,
    last_seen_at: 1000,
    connected_at: null,
    ...overrides,
  };
}
